interfaces/builtin/posix_mq.go

```diff
@@ -0,0 +1,291 @@
+		BaseDeclarationSlots: posixMQBaseDeclarationSlots,
+		BaseDeclarationPlugs: posixMQBaseDeclarationPlugs,
+	}
+}
+
+func (iface *posixMQInterface) Name() string {
+	return "posix-mq"
+}
+
+func (iface *posixMQInterface) checkPosixMQAppArmorSupport() error {
+	if apparmor_sandbox.ProbedLevel() == apparmor_sandbox.Unsupported {
+		/* AppArmor is not supported at all; no need to add rules */
+		return nil
+	}
+	if features, err := apparmor_sandbox.ParserFeatures(); err == nil {
+		if !strutil.ListContains(features, "mqueue") {
+			return fmt.Errorf("AppArmor does not support POSIX message queues - cannot setup or connect interfaces")
+		}
+	} else {
+		return err
+	}
+	return nil
+}
+
+func (iface *posixMQInterface) isValidPermission(perm string) bool {
+	for _, validPerm := range posixMQPlugPermissions {
+		if perm == validPerm {
+			return true
+		}
+	}
+	return false
+}
+
+func (iface *posixMQInterface) validatePermissionList(perms []string, name string) error {
+	for _, perm := range perms {
+		if !iface.isValidPermission(perm) {
+			return fmt.Errorf("posix-mq slot %s permission \"%s\" not valid, must be one of %v", name, perm, posixMQPlugPermissions)
+		}
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) getPermissions(attrs interfaces.Attrer, name string) ([]string, error) {
+	var perms []string
+
+	if err := attrs.Attr("permissions", &perms); err != nil {
+		/* If the permissions have not been specified, use the defaults */
+		perms = posixMQDefaultPlugPermissions
+	}
+
+	if err := iface.validatePermissionList(perms, name); err != nil {
+		return nil, err
+	}
+
+	return perms, nil
+}
+
+func (iface *posixMQInterface) getPath(attrs interfaces.Attrer, name string) (string, error) {
+	if pathAttr, isSet := attrs.Lookup("path"); isSet {
+		if path, ok := pathAttr.(string); ok {
+			return path, nil
+		}
+		return "", fmt.Errorf(`posix-mq slot %s "path" attribute must be a string, not %v`, name, pathAttr)
+	}
+	return "", fmt.Errorf(`posix-mq slot %s has missing "path" attribute`, name)
+}
+
+func (iface *posixMQInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	if err := iface.checkPosixMQAppArmorSupport(); err != nil {
+		return err
+	}
+
+	/* Ensure that the given permissions are valid */
+	if _, err := iface.getPermissions(plug, plug.Name); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	if err := iface.checkPosixMQAppArmorSupport(); err != nil {
+		return err
+	}
+
+	/* Ensure that the given permissions are valid */
+	if _, err := iface.getPermissions(slot, slot.Name); err != nil {
+		return err
+	}
+
+	/* Ensure that the given path has a valid POSIX MQ name */
+	if path, err := iface.getPath(slot, slot.Name); err == nil {
+		if posixMQNamePattern.MatchString(path) {
+			return nil
+		}
+		return fmt.Errorf("posix-mq path must conform to the POSIX message queue name specifications (see `man mq_overview`)")
+	} else {
+		return err
+	}
```

Comment:
Can this path also please be validated using `apparmor_sandbox.ValidateNoAppArmorRegexp()`? In general I think it would be useful to structure the checks similarly to say how the currently-in-development `pkcs11` interface is over in https://github.com/snapcore/snapd/pull/11540